drivers/scsi/ses.c

```diff
@@ -766,10 +788,27 @@ static int ses_intf_add(struct device *cdev,
 
 	/* see if there are any devices matching before
 	 * we found the enclosure */
-	shost_for_each_device(tmp_sdev, sdev->host) {
-		if (tmp_sdev->lun != 0 || scsi_device_enclosure(tmp_sdev))
-			continue;
-		ses_match_to_enclosure(edev, tmp_sdev, 0);
+	if (scsi_is_ahciem(sdev)) {
+		struct Scsi_Host *shost = sdev->host;
+		struct ahciem_enclosure *enc = (struct ahciem_enclosure *)&shost->hostdata[0];
+		struct ata_host *host = enc->host;
+
+		for (i = 0; i < host->n_ports; i++) {
+			struct ata_port *ap = host->ports[i];
+
+			if (!ap)
+				continue;
+
+			shost_for_each_device(tmp_sdev, ap->scsi_host) {
+				ses_match_to_enclosure(edev, tmp_sdev, 0);
```

Comment:
Here it is totally upside-down -- SES already reports you bus numbers for each slot.  You should traverse through the slots and for each one look for the devices.

Reply:
Oh insightful! That lets me make the ahciem_enclosure structure private again, too!